README.md
# Portable Object Compiler++ #

### Overview ###

This is the **Portable Object Compiler++**, a branch of David Stes' *Portable
Object Compiler*.

It provides a compiler for *Objective-C*, which includes support for novel
functionality like *Blocks* (which are an implementation of closures or lambda
expressions).

A standard library based on *ICpak 101* is included; this is called
**ObjPak++**. This provides a powerful set of container classes which work
similarly to those provided in the *SmallTalk-80* standard library.

A variety of other libraries are also available; one of these is **CAkit**, the
*Computer Algebra Kit*, which provides classes allowing multiple precision
arithmetic and polynomial mathematics.

**CAkit** and an *Objective-C* interpreter and debugger known as VICI are found
in the PC++-Extras repository.

### Compiling ###

Instructions are available in `doc/compile` detailing compilation on various
platforms.

Before you can build the Portable Object Compiler, you will need to build and
install the bootstrap package. This is available in the **Releases** section of
the GitHub page.


## Original Readme.txt ##

The original Readme.txt follows below:


Portable Object Compiler (c) 1997,98,99,2000,01,02,03,04.  All Rights Reserved.


GETTING HELP
------------

For help and reporting bugs : send email to objc-users@lists.sourceforge.net
(to subscribe : send email to objc-users-request@lists.sourceforge.net)

INSTALL
-------

Read the Install.txt file.

IMPORTANT : if this is a first-time installation, you have to start by installing the bootstrap package; do not install this package until you have installed the bootstrap package.

For every release, there is an objc-x.y.z.tar.gz package and an objc-x.y.z-bootstrap.tar.gz package.

LICENSE
-------

Read the License.txt file (the "GNU Library General Public License").

The license file explains the difference between,

 (a) "software using the Portable Object Compiler", and 
 (b) "software based upon the Portable Object Compiler".


LATEST VERSION
--------------

The latest bootstrap compilers and sources of the Portable Object Compiler
are available at:

	http://sunsite.unc.edu/pub/Linux/devel/lang/objc

Or from my home page at:

	http://users.pandora.be/stes/

WHAT'S INCLUDED
---------------

The "objc" distribution consists of the following parts :

	0. "VICI", interactive Objective-C debugger and interpreter

	1. "objc", the actual compiler

	Consists of driver script ("objc") and binary ("objc1").
	"objc1" is a precompiler for C.
	The binary "postlink" is used
	for preparing tables for runtime initialization.

	2. "objcrt", the Objective C Runtime Library for "objc"

	This is a modern, portable, "all C" Objective C runtime
	There's absolutely no assembly language needed, nor is there
	a dependency on stack layout, register allocation conventions etc.

	Most other runtimes, were derived from pre-"all C"-Stepstone 
	runtimes, so this is an important point of difference.

	3. "objpak", the "Object Pak" Objective C Collection Classes

	Simple, powerful set of classes.  Very portable.
	Can be used to develop with different compilers, then
	port with 'objc'.  Compatible with ICpak101, the collection class
	library described in Brad Cox book, and used by existing
	Objective-C packages that were developed with Stepstone objc.

	4. "cakit", the "Computer Algebra Kit" Objective C classes

	Small, concise, interface to a large set of classes for
	polynomial computing and arbitrary precision integer arithmetic.

	5. Source of compiler and "oclib", a set of Objective C classes for
        parsing C (and the Objective C extensions to C)
  
"objc" and "vici" are written in Objective C itself.  It's all
written in portable Objective C : any Objective C compiler
should be able to compile the "objc" compiler.  From time to time,
we try a configure with other Objective-C compilers, to ensure that our
sources are still compatible with e.g. Stepstone Objective C compiler.

NOT included is a C compiler : in order to be able to use "objc", you will
need to elsewhere find a compatible C compiler.  This should be no problem
since one of our goals - see below - is to make "objc" work with as many C
compilers as possible.  On UNIX systems, the (free) "gcc" compiler is a
possibility (not needed, but possible).  Another possibility, is to use
the (free) "lcc" compiler or the "tcc" TDF compiler system.


NICE FEATURES
-------------

	1.Easy to install or to modify.

	2.Works on many systems with the native cc, debugger, profiler etc.
	(Unix, Windows, Macintosh, Beos, OpenVMS etc. see Platforms.txt file)

	2a. Option for reference counted memory management (-refcnt).
	This uses the native malloc(), free() etc. but the compiler generates
	statements for keeping track of references.
	(Tested on a few platforms, such as IRIX 5.2 with the SGI malloc).

	3.Built-in possibility of tracing Objective C messages. (OBJCRTMSG)

	4.Straightforward "C" messenger; "inline cache" messenger.
	Forwarding C messenger (to support -doesNotUnderstand:).
	
	5.All classes get a +initialize message at start-up, rather than
	'each class receives a +initialize before it receives its first
	message'.

	6.Some support for translating Objective-C to Smalltalk (-st80 option)

	7.Automatic archiver.  Compiler generates code for classes to save
	and load objects to and from disk (for all instance variables of
	type "id").

	8.Option for Garbage Collection (using Boehm gc package).  Tested
	on some UNIXes and WIN32.   Option for reference counted memory
	management (doesn't require Boehm).

	9.Exception handling scheme (using Objective-C Blocks) that allows
	to specify a default handler to be executed.

	10.Supports dynamically loading Objective-C modules on Windows,
	FreeBSD, HP-UX, Linux, IRIX, Digital Unix etc.

	11.Has a switch for double indirection for Object identifiers (id as
	a handle instead of a pointer). (-become: method)

	12.Supports forwarding messages (-doesNotUnderstand: method)

        13.Support for Embedded SQL in Objective-C (Informix only for now)  

	14.Great system for experimentation with your own additions/extensions
	to Objective C !


WANT TO HELP?
-------------

If you like "objc", feel free to send e-mail to, "stes@pandora.be".
I'm especially interested in bug-fixes, of course.  And in ports
to machines that I don't have access to (e.g. CP/M, Atari).



doc/Changes.txt

3.2.12  * port to Oracle Solaris 11 with Sun C 5.12 SunOS_i386 2011/11/16

        * support thread-local storage C keyword __thread
	specifically have to support due to Solaris <strings.h> following:

	extern __thread char *__strdupa_str;
	extern __thread size_t __strdupa_len;

        * back out __builtin_alloca as gnu builtin for Sun Studio which
          defines it in header as regular function

	* port to Slackware 14.1 with gcc 4.8.2 and flex 2.5.37
	add __builtin_bswap32 and __builtin_bswap64 builtin functions

	* port to FreeBSD 9.2 with clang 3.3 and gcc 4.2.1 (flex 2.5.4 byacc)
	FreeBSD 9.2 has /usr/include/stdlib.h : _Noreturn void exit(int);
	(in their __ISO_C_VISIBLE >= 2011 mode, 'greater than C99 expose C1x')
	stdlib.h : int abs(int) __pure2;
	stdlib.h : void abort2(const char*,int,void**) __dead2;
	_Noreturn, __pure2 and __dead2 are macros __attribute((__noreturn__))
	__attribute__((__const__)) and __attribute__((__noreturn__))

 	NOTE: before
	 byacc: 6 shift/reduce conflicts, 48 reduce/reduce conflicts.
	AFTER hack to support the _Noreturn attribute
 	byacc: 8 shift/reduce conflicts, 49 reduce/reduce conflicts.

	* fix for FreeBSD yacc __unused static char const yyrcsid[]="..."
	where __unused expands to __attribute__((__unused__))

	NOTE: after fix
	byacc: 8 shift/reduce conflicts, 49 reduce/reduce conflicts.

	* add __builtin_strncpy for CentOS 6.5 in /usr/include/string.h

3.2.11	* fix for Ubuntu header file, reported by adriana.koncsag@yahoo.com
          __attribute problems, extern_always_inline problem in cdefs.h

	* fix for bootSrc on SunOS 11 (pragma tab issue in util.m)

	* add errno.h to util directory

	* some changes for -refcnt, isvoid problem with __attribute__(()) void
	  not correctly recognized as void in -refcnt case
          try to filter out __attribute__ stuff using isgnuattrib

3.2.10	* add patch Finn Wilcox finnw
	  ocstring cannot concat: or at:insert: itself - ID: 1640788

	* port to OpenSolaris 09/06 with Sun C 5.10 SunOS_i386 2009/06/03

	* updated email address in indent-patch

3.2.9	* add OpenSolaris (solaris 2.11) to Platforms

	* change __builtin_va_alist to builtinvar for OpenSolaris

3.2.8   * add __builtin_expect, __builtin_strchr and more garbage for gcc 3.4.6
	  __builtin_cmp (GCC developers seem to believe they're using PASCAL)

        * regenerated configure with autoconf 2.6

	* upgraded config.guess and config.sub to automake 1.9

	* add NIOS2 sopc to platforms  (port by lionnel@ipricot.com)

	* rebuilt bootstrap package

	* upgraded nm objc1 | postlink >util/_objc1.c 

3.2.7	* fix for -builtintype and -builtinfunction broken
          (pragma OCbuiltInFctn and OCbuiltInType still worked)

	* port to IA-64 hpux 11.23 with HP aC++/ANSI C (add to platforms)

	* add -hpux flag with some builtins: __va_list__, __va_arg__,
          __va_start__, __fpreg and __float80

3.2.6   * add patch of yerfino@delsatgroup.com for ordcltn -includes:

        * some filer tweaking

3.2.5   * add message string to TypeInconsistency class

	* increased Filer version number

3.2.4	* add exception subclasses BadVersion, TypeInconsistency, UnknownType

	* add LCC patches for MAKEFILE  + manual addition LCC

3.2.3	* apply yacc.ym + CompoundStmt patch of Mike for further C99 support

	* continue work on Exception (add instance variable 'resumeHandler')

3.2.2	* add subclass "OutOfMemory" (inherits from Exception) to objcrt

	* add subclass "OutOfBounds" (inherits from Exception) to objpak

	* add subclass "NotFound" (inherits from Exception) to objpak

	* add lex patch of Mike for partial C99 support
          (the yacc patch introduces some problems for the moment)

	* updated file Intro.txt to show Exception classes

3.2.1   * new class Exception and new message -on:do: for Block class

          the -on:do: message is to keep in line with Ansi Smalltalk
          developments ...

	  See the documentation on -on:do: (in the Block class for more info)

3.1.33  * temporarily undo 3.1.31 lexfiltr by setting the flag in the driver
          to no by default (there was a problem reported on LCC with this)
	
	* fix minor problem with show message with -refcnt
	  this problem was basically due to the fact that the runtime is
	  compiled without -refcnt, while Object.m is compiled with -refcnt,
	  and one should not mix modules compiled with/without -refcnt

	* add documentation for includesAllOf:, includesAnyOf: and -addAll:
	  (thanks to Ralph Moeritz for pointing out that this was missing)

	* add note on -ansic to the Lcc.txt readme file

	* changed email address of the mailing list

3.1.32  * throw out change in yacc.ym for -refcnt as it breaks the compile
          with GCC 3.2 and -O2; the -O2 option enables some GNU braced
          expressions in the header files and GNU braced expressions have some
          restrictions 

3.1.31  * update config.guess, config.sub and install-sh to automake 1.5

	* add updated lexfiltr.m from Lee Willis

	* add notes on using the debugger to Visual.txt

	* add notes on using the Visual C++ GUI to Visual.txt

        * add a test to the configure script to test for support for #line
          and # in .i files;  this is because there is a bug recently
          introduced in gcc 3.1 where support for #line has been removed

3.1.30  * add patch from Lee Willis for lex.lm DosAsmStmt which adds an extra
          case for the patch by Ronald Pijnacker/JVdGoor

        * support for duplicate typedefs, patch by lee-willis@attbi.com

	  POC now supports duplicate typedefs like icc or msvc visual C
		typedef int FOO;
		typedef int FOO;
	  is accepted now as input (see src/objc/lexfiltr.m)

	  this is enabled with the option -duptypedefs or -msdos

3.1.29  * support for Intel icc for Linux (Version 6.0   Build 020312Z)
         (add __ALIGNOF__ builtin fun)

         CC=icc -w
         CPP=icc -E
         export CC CPP
	 ./configure 

3.1.28  * fix stack tracing print feature which was broken due to change in

objc/objc1.m

/*
 * Copyright (c) 1998,99,2000,01,02,03,04,06,08,09,14 David Stes
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Library General Public License as published
 * by the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#define OBJCVERSION "3.2.12"
#define OBJC1VERSION "objc3.2.12"

#include <string.h> /* strtok */
#include <stdlib.h> /* getenv */
#include <assert.h>
#ifndef __OBJECT_INCLUDED__
#define __OBJECT_INCLUDED__
#include <stdio.h>  /* FILE */
#include "Object.h" /* Stepstone Object.h assumes #import */
#endif
#include <ocstring.h>
#include <ordcltn.h>
#include <sequence.h>
#include "node.h"
#include "trlunit.h"
#include "options.h"
#include "util.h"
#include "symbol.h"

void printversion (void)
{
    printf ("%s\n", OBJCVERSION);
    exit (0);
}

void printcopyright (void)
{
    printf ("Portable Object Compiler++ %s (c) 1997-2015.\n", OBJCVERSION);
    printf ("Use is subject to licence terms.\n");
}

void unknownoption (char * arg)
{
    STR msg  = "%s: unknown option %s\n";
    STR name = (o_cplus) ? "objcpls1" : "objc1";
    fprintf (stderr, msg, name, arg);
    exit (1);
}

void badarg (id option, id arg)
{
    STR msg  = "%s: illegal argument %s for %s\n";
    STR name = (o_cplus) ? "objcpls1" : "objc1";
    fprintf (stderr, msg, name, [arg str], [option str]);
    exit (1);
}

BOOL isoption (char * s) { return s != NULL && s[0] == '-'; }

id optionstok (id aCltn, char * s)
{
    char * p;
    char * delims = " \t\n\r";
    id buffer     = [String str:s];

    p = strtok ([buffer str], delims);

    while (p != NULL)
    {
        [aCltn add:[String str:p]];
        p = strtok (NULL, delims);
    }

    return aCltn;
}

id optsfromfile (id aCltn, FILE * f)
{
    char buf[BUFSIZ + 1];

    while (!feof (f))
    {
        if (fgets (buf, BUFSIZ, f))
        {
            /* this works 'cause strtok also deletes \n */
            optionstok (aCltn, buf);
        }
    }

    return aCltn;
}

/* this is mostly for MS-DOS cmd line length limitations */

id optsnamed (id aCltn, STR s)
{
    STR t;
    FILE * f;

    /* maybe there's such an environment variable */
    if ((t = getenv (s)))
        return optionstok (aCltn, t);

    /* maybe there's such a file */
    if ((f = fopen (s, "r")))
    {
        aCltn = optsfromfile (aCltn, f);
        fclose (f);
        return aCltn;
    }
    /* nope */
    return nil;
}

id cmdlineopts (int argc, char * argv[])
{
    int i;
    id aCltn = [OrdCltn new];

    for (i = 1; i < argc; i++)
    {
        [aCltn add:[String str:argv[i]]];
    }

    return aCltn;
}

/*
 * Filename for first line tag output (source filename)
 * (which we want to be file.m instead of file.P)
 */

int setfilename (id option, id args)
{
    id arg   = [args next];
    char * t = [arg str];

    if (isoption (t))
        badarg (option, arg);
    o_srcfilename = t;

    return 0;
}

int setinitcall (id option, id args)
{
    id arg   = [args next];
    char * t = [arg str];

    if (isoption (t))
        badarg (option, arg);
    o_initcall = t;

    return 0;
}

int setbrowsedir (id option, id args)
{
    id arg   = [args next];
    char * t = [arg str];

    if (isoption (t))
        badarg (option, arg);
    o_browsedir = t;

    return 0;
}

int setmainfun (id option, id args)
{
    id arg = [args next];
    if (isoption ([arg str]))
        badarg (option, arg);
    o_mainfun = [arg str];
    return 0;
}

int setbind (id option, id args)
{
    id arg = [args next];
    if (isoption ([arg str]))
        badarg (option, arg);
    o_bind = [arg str];
    return 0;
}

/*
 * Make line max settable (works only for identifiers now).
 */

int setlinemax (id option, id args)
{
    id arg = [args next];
    if (isoption ([arg str]))
        badarg (option, arg);
    o_linemax = [arg asInt];
    return 0;
}

/*
 * @ filename is understood to take options from "filename"
 * Maybe we should also allow @filename (without space)
 */

void setoptions (id aCltn);

int setoptsfromfile (id option, id args)
{
    id arg = [args next];

    if (isoption ([arg str]))
        badarg (option, arg);
    setoptions (optsnamed ([OrdCltn new], [arg str]));
    return 0;
}

/*
 * Builtin functions (a la __builtin_isfloat).
 */

int addbuiltintype (id option, id args)
{
    id arg = [args next];
    if (isoption ([arg str]))
        badarg (option, arg);
    definebuiltintype ([arg str]);
    return 0;
}

int addbuiltinfun (id option, id args)
{
    id arg = [args next];
    if (isoption ([arg str]))
        badarg (option, arg);
    definebuiltinfun ([arg str]);
    return 0;
}

void setoptions (id aCltn)
{
    id args, s;
    int filecount    = 0;   /* optional inFile and outFile */
    BOOL checkoption = YES; /* YES if filename can't begin with a dash */

    args = [aCltn eachElement];

    while ((s = [args next]))
    {
        char * t = [s str];
        if (!strcmp (t, "-quiet"))
        {
            ++o_quiet;
        }
        else if (!strcmp (t, "-z"))
        {
            ++o_outputcode;
        }
        else if (!strcmp (t, "@"))
        {
            setoptsfromfile (s, args);
        }
        else if (!strcmp (t, "-u"))
        {
            o_buffered = 0;
        }
        else if (!strcmp (t, "-version"))
        {
            o_version++;
        }
        else if (!strcmp (t, "-otb"))
        {
            o_otb++;
            o_shareddata = 0;
        }
        else if (!strcmp (t, "-gnu"))
        {
            o_gnu++;
        }
        else if (!strcmp (t, "-vms"))
        {
            o_vms++;
        }
        else if (!strcmp (t, "-hpux"))
        {
            o_hpux++;
        }
        else if (!strcmp (t, "-refcnt"))
        {
            o_refcnt++;
            o_selfassign = 0; /* because receiver of a message is not counted */
        }
        else if (!strcmp (t, "-noasm"))
        {
            o_enableasm = 0;
        }
        else if (!strcmp (t, "-nolonglong"))
        {
            o_llkeyw = 0;
        }
        else if (!strcmp (t, "-msdos"))
        {
#if 0
 /* there are still some problems with this code on LCC */
 /* temporarily disable it (with the goal to enable this later again) */
      o_duptypedefs++;
#endif
            o_msdos++;
        }
        else if (!strcmp (t, "-watcom"))
        {
            o_watcom++;
        }
        else if (!strcmp (t, "-comments"))
        {
            o_comments++;
        }
        else if (!strcmp (t, "-duptypedefs"))
        {
            o_duptypedefs++;
        }
        else if (!strcmp (t, "-ibmvac"))
        {
            o_ibmvac++;
        }
        else if (!strcmp (t, "-noSelTbl"))
        {
            o_seltranslation = 0;
        }
        else if (!strcmp (t, "-noCategories"))
        {
            o_categories = 0;
        }
        else if (!strcmp (t, "-noBlocks"))
        {
            o_blocks = 0;
        }
        else if (!strcmp (t, "-st80"))
        {
            o_st80++;
            o_nolinetags++;
            o_outputcode = 0; /* no C code */
        }
        else if (!strcmp (t, "-fwd"))
        {
            o_fwd++;
        }
        else if (!strcmp (t, "-noSelPtr"))
        {
            o_selptr = 0;
        }
        else if (!strcmp (t, "-noFwd"))
        {
            o_fwd = 0;
        }
        else if (!strcmp (t, "-noCache"))
        {
            o_cache = 0;
        }
        else if (!strcmp (t, "-noFiler"))
        {
            o_filer = 0;
        }
        else if (!strcmp (t, "-noSelfAssign"))
        {
            o_selfassign = 0;
        }
        else if (!strcmp (t, "-noNilRcvr"))
        {
            o_nilrcvr = 0;
        }
        else if (!strcmp (t, "-objc"))
        {
            o_gencode = 0;
        }
        else if (!strcmp (t, "-cplus"))
        {
            o_cplus++;
        }
        else if (!strcmp (t, "-inlinecache"))
        {
            o_inlinecache++;
        }
        else if (!strcmp (t, "-refBind"))
        {
            o_refbind++;
        }
        else if (!strcmp (t, "-export"))
        {
            setbind (s, args);
        }
        else if (!strcmp (t, "-dllexport"))
        {
            o_bind = "__declspec(dllexport)";
        }
        else if (!strcmp (t, "-traditional"))
        { /* ignore */
        }
        else if (!strcmp (t, "-filename"))
        {
            setfilename (s, args);
        }
        else if (!strcmp (t, "-longlinetag"))
        { /* compatib. */
        }
        else if (!strcmp (t, "-shortTags"))
        {
            o_tagformat = "# %d \"%s\"\n";
        }
        else if (!strcmp (t, "-init"))
        {
            setinitcall (s, args);
        }
        else if (!strcmp (t, "-browse"))
        {
            setbrowsedir (s, args);
            o_nolinetags++;
            o_outputcode = 0; /* no C code */
        }
        else if (!strcmp (t, "-noShared"))
        {
            o_shareddata = 0;
        }
        else if (!strcmp (t, "-main"))
        {
            setmainfun (s, args);
        }
        else if (!strcmp (t, "-linemax"))
        {
            setlinemax (s, args);
        }
        else if (!strcmp (t, "-builtinfunction"))
        {
            addbuiltinfun (s, args);
        }
        else if (!strcmp (t, "-builtintype"))
        {
            addbuiltintype (s, args);
        }
        else if (!strcmp (t, "-nolinetags"))
        {
            o_nolinetags++;
        }
        else if (!strcmp (t, "-oneperfile"))
        {
            o_oneperfile++;
        }
        else if (!strcmp (t, "-w"))
        {
            o_warnings = 0;
        }
        else if (!strcmp (t, "-wClassUsedAsType"))
        {
            o_warnclasstype = 0;
        }
        else if (!strcmp (t, "-wMissing"))
        {
            o_warnmissingmethods = 0;
        }
        else if (!strcmp (t, "-wTypeConflict"))
        {
            o_warntypeconflict = 0;
        }
        else if (!strcmp (t, "-wLocalInstance"))
        {
            o_warnlocalnst = 0;
        }
        else if (!strcmp (t, "-wUndefinedMethod"))
        {
            o_warnundefined = 0;
        }
        else if (!strcmp (t, "-wInterfaceNotFound"))
        {
            o_warnnotfound = 0;
        }
        else if (!strcmp (t, "-WLex"))
        {
            o_warnlex++;
        }
        else if (!strcmp (t, "-WFwd"))
        {
            o_warnfwd++;
        }
        else if (!strcmp (t, "-postlink"))
        {
            o_postlink++;
        }
        else if (!strcmp (t, "-checkbindfunction"))
        {
            o_checkbind++;
        }
        else if (!strcmp (t, "-debugInfo"))
        {
            o_debuginfo++;
        }
        else if (!strcmp (t, "-ppi"))
        {
            o_ppi++;
        }
        else if (!strcmp (t, "-"))
        {
            checkoption = 0;
        }
        else if (checkoption && isoption (t))
        {
            unknownoption (t);
        }
        else if (filecount == 0)
        {
            o_infile = t;
            filecount++;
        }
        else if (filecount == 1)
        {
            o_outfile = t;
            filecount++;
        }
        else
        {
            unknownoption (t);
        }
    }
}

/*
 * Opening and closing input file.  We're not dealing with wrapping around
 * files (include)  (because we're processing cpp output)
 */

void openinfile (void)
{
    if (o_infile)
    {
        yyin = openfile (o_infile, "r");
    }
    else
    {
        /* stdin */
    }
}

void setfirstlinetag (void)
{
    /* initialize linetag mechanism */

    inlineno = 1;
    if (o_srcfilename)
    {
        infilename = [String str:o_srcfilename];
    }
    else
    {
        if (o_infile)
        {
            infilename = [String str:o_infile];
        }
        else
        {
            infilename = [String str:"-=stdin=-"];
        }
    }

    if (!o_nolinetags)
    {
        gl (inlineno, [infilename str]);
    }

    assert (trlunit != nil);
    [trlunit setmodversion:OBJC1VERSION];
    [trlunit setmodname:[infilename str]];

    definebuiltinvar ("__func__");

    if (o_vms)
    {
        definebuiltinfun ("__VA_START_BUILTIN");
        definebuiltinfun ("__IS_X_FLOAT_BUILTIN");
        definebuiltintype ("__int64");
    }

    if (o_hpux)
    {
        definebuiltinfun ("__va_start__");
        definebuiltinfun ("__va_arg__");
        definebuiltintype ("__va_list__");
        definebuiltintype ("__fpreg");
        definebuiltintype ("__float80");
    }

    if (o_gnu)
    {
        definebuiltinfun ("__builtin_saveregs");
        definebuiltinfun ("__builtin_classify_type");
        definebuiltinfun ("__builtin_isfloat");
        definebuiltinfun ("__builtin_classof");
        definebuiltinfun ("__alignof__");
        definebuiltinfun ("__ALIGNOF__"); /* Intel icc 6.0 header */
        definebuiltinfun ("__builtin_alignof");
        definebuiltinfun ("__builtin_next_arg");
        /*  definebuiltinfun("__builtin_alloca"); Sun Solaris 11.1 regular
         * function */
        definebuiltinfun ("__builtin_memcpy");
        definebuiltinfun ("__builtin_memset");
        definebuiltinfun ("__builtin_fabs");
        definebuiltinfun ("__builtin_fabsl");
        definebuiltinfun ("__builtin_fabsf");
        definebuiltinfun ("__builtin_constant_p");
        definebuiltinvar ("__PRETTY_FUNCTION__");
        definebuiltinvar ("_VA_FP_SAVE_AREA");
        definebuiltintype ("__builtin_va_list"); /* IA-64 */
        definebuiltinfun ("__builtin_stdarg_start");
        definebuiltinfun ("__builtin_va_end");
        definebuiltinfun ("__builtin_va_arg");
        definebuiltinfun ("__builtin_va_start");
        definebuiltinfun ("__builtin_expect");
        definebuiltinfun ("__builtin_strchr");
        definebuiltinfun ("__builtin_strcmp");
        definebuiltinfun ("__builtin_strncmp");
        definebuiltinfun ("__builtin_strncpy"); /* CentOS 6.5 gcc 4.4.7 */

        /* gcc 4.5 for ubuntu 11.04 port */
        definebuiltinfun ("__builtin_strlen");
        definebuiltinfun ("__builtin_object_size");
        definebuiltinfun ("__builtin_va_arg_pack");
        definebuiltinfun ("__builtin___sprintf_chk");
        definebuiltinfun ("__builtin___vsprintf_chk");
        definebuiltinfun ("__builtin___snprintf_chk");
        definebuiltinfun ("__builtin___vsnprintf_chk");
        definebuiltinfun ("__builtin___strcpy_chk");
        definebuiltinfun ("__builtin___strncpy_chk");
        definebuiltinfun ("__builtin___strcat_chk");
        definebuiltinfun ("__builtin___strncat_chk");
        definebuiltinfun ("__builtin___memmove_chk");
        definebuiltinfun ("__builtin___memset_chk");
        definebuiltinfun ("__builtin___memcpy_chk");

        /* Slackware 14.1 port gcc 4.8.2  */
        definebuiltinfun ("__builtin_bswap32");
        definebuiltinfun ("__builtin_bswap64");

        /* the following are not really gnu, they're SUN headers */
        definebuiltinvar ("__builtin_va_alist");
        definebuiltinfun ("__builtin_va_arg_incr");
    }
}

void closeinfile (void)
{
    if (o_infile)
    {
        fclose (yyin);
    }
    else
    {
        /* stdin */
    }
}

void openoutfile (void)
{
    if (o_outfile)
    {
        gfile = openfile (o_outfile, "w");
    }
    else
    {
        gfile = stdout; /* direct output to this stream */
    }

    /* for debugging output (-u unbuf option) */
    if (!o_buffered)
        setbuf (gfile, NULL);
}

void closeoutfile (void)
{
    if (o_outfile)
        fclose (gfile);
}

int main (int argc, char * argv[])
{
    /* define early since setoptions() modifies this object */
    o_mainfun = "main";
    trlunit   = [TranslationUnit new];

    defoptions ();
    setoptions (cmdlineopts (argc, argv));

    if (o_version)
        printversion ();
    if (!o_quiet)
        printcopyright ();
    openinfile ();
    openoutfile (); /* redirect stdout */

    setfirstlinetag ();

    if (o_outputcode)
        [trlunit prologue];
    if (yyparse ())
        exit (1);
    if (o_outputcode)
        [trlunit epilogue];

    closeinfile ();
    closeoutfile ();

    if (o_browsedir)
        [trlunit browse];

    exit (exitstatus);
    return 0;
}
